library/network/test/y2firewall/firewalld/api/zones_test.rb

```diff
@@ -0,0 +1,213 @@
+        .and_return(false)
+      expect(subject.masquerade_enabled?("public")).to eql(false)
+    end
+
+    it "returns true if the zone has masquerade enabled" do
+      allow(api).to receive(:query_command)
+        .with("--zone=external", "--query-masquerade", permanent: api.permanent?)
+        .and_return(true)
+
+      expect(subject.masquerade_enabled?("external")).to eql(true)
+    end
+  end
+
+  describe "#list_ports" do
+    it "returns the list of ports opened by the zone" do
+      allow(api).to receive(:string_command)
+        .with("--zone=test", "--list-ports", permanent: api.permanent?)
+        .and_return("80/tcp 443/tcp")
+
+      expect(api.list_ports("test")).to eql(["80/tcp", "443/tcp"])
+    end
+  end
+
+  describe "#list_protocols" do
+    it "returns the list of protocols opened by the zone" do
+      allow(api).to receive(:string_command)
+        .with("--zone=test", "--list-protocols", permanent: api.permanent?)
+        .and_return("egp gre")
+
+      expect(api.list_protocols("test")).to eql(["egp", "gre"])
+    end
+  end
+
+  describe "#remove_port" do
+    it "removes the given port from the zone configured ports" do
+      expect(api).to receive(:modify_command)
+        .with("--zone=test", "--remove-port=80/tcp", permanent: api.permanent?)
+
+      api.remove_port("test", "80/tcp")
+    end
+  end
+
+  describe "#add_port" do
+    it "adds the given port to the zone configured ports" do
+      expect(api).to receive(:modify_command)
+        .with("--zone=test", "--add-port=80/tcp", permanent: api.permanent?)
+
+      api.add_port("test", "80/tcp")
+    end
+  end
+
+  describe "#port_enabled?" do
+    it "returns false if the port is not allowed by the zone" do
+      allow(api).to receive(:query_command)
+        .with("--zone=public", "--query-port=80/tcp", permanent: !api.offline?)
+        .and_return(false)
+      expect(subject.port_enabled?("public", "80/tcp")).to eql(false)
+    end
+
+    it "returns true if the port is allowed by the zone" do
+      allow(api).to receive(:query_command)
+        .with("--zone=public", "--query-port=22/tcp", permanent: !api.offline?)
+        .and_return(true)
+      expect(subject.port_enabled?("public", "22/tcp")).to eql(true)
+    end
+  end
+
+  describe "#port_enabled?" do
+    it "returns false if the port is not allowed by the zone" do
+      allow(api).to receive(:query_command)
+        .with("--zone=public", "--query-port=80/tcp", permanent: api.permanent?)
+        .and_return(false)
+      expect(subject.port_enabled?("public", "80/tcp")).to eql(false)
+    end
+
+    it "returns true if the port is allowed by the zone" do
+      allow(api).to receive(:query_command)
+        .with("--zone=public", "--query-port=22/tcp", permanent: api.permanent?)
+        .and_return(true)
+      expect(subject.port_enabled?("public", "22/tcp")).to eql(true)
+    end
+  end
+
+  describe "#protocol_enabled?" do
```

Comment:
This is just a matter of taste, but using a before block could save some repetition:

```ruby
describe "#protocol_enabled?" do
  before do
    allow(api).to receive(:query_command)
      .with("--zone=public", "--query-protocol=igmp", permanent: api.permanent?)
      .and_return(allowed?)
  end

  context "when the protocol is not allowed by the zone" do
    let(:allowed?) { false }

    it "returns false" do
      expect(subject.protocol_enabled?("public", "igmp")).to eql(false)
    end
  end

  context "when the protocol is allowed by the zone" do
    let(:allowed?) { true }

    it "returns true" do
      expect(subject.protocol_enabled?("public", "igmp")).to eql(true)
    end
  end
end
```

Just a suggestion, but it is ok as it is right now.

Reply:
OK, will move next time, it save the repetition but needs to add the contexts and lets per each, so added the explicit allowements :see_no_evil:

Reply:
Yes, that's the point. In my opinion the intention is more clear. You set the "preconditions" outside of the test and the it block only contains the relevant part. But, I've said, it is a matter of taste.